File: examples/legal_moves.py
import spooky_chess as sc


def side_name(color: int) -> str:
    return "White" if color == sc.WHITE else "Black"


def main() -> None:
    game = sc.Game.standard()
    assert game.make_move_from_san("e4")
    assert game.make_move_from_san("e5")

    print("Position after 1.e4 e5")
    print(f"Turn: {side_name(game.turn())}")
    print(f"In check: {game.is_check()}")

    legal_rows = sorted((game.move_to_san(move_), move_.to_lan()) for move_ in game.legal_moves())
    print(f"{len(legal_rows)} legal moves:")
    for san, lan in legal_rows:
        print(f"  {san:<6} {lan}")

    knight = sc.Position.from_algebraic("g1")
    knight_rows = sorted(
        (game.move_to_san(move_), move_.to_lan()) for move_ in game.legal_moves_for_position(knight.col(), knight.row())
    )
    print()
    print("Moves from g1:")
    for san, lan in knight_rows:
        print(f"  {san:<6} {lan}")


if __name__ == "__main__":
    main()

File: examples/legal_moves.rs
#![feature(generic_const_exprs)]
#![allow(incomplete_features)]

use std::error::Error;

use spooky_chess::color::Color;
use spooky_chess::game::StandardGame;
use spooky_chess::position::Position;

fn side_name(color: Color) -> &'static str {
    match color {
        Color::White => "White",
        Color::Black => "Black",
    }
}

fn main() -> Result<(), Box<dyn Error>> {
    let mut game = StandardGame::standard();
    assert_eq!(game.make_move_from_san("e4"), Ok(true));
    assert_eq!(game.make_move_from_san("e5"), Ok(true));

    println!("Position after 1.e4 e5");
    println!("Turn: {}", side_name(game.turn()));
    println!("In check: {}", game.is_check());

    let legal_moves = game.legal_moves();
    let mut legal_rows: Vec<_> = legal_moves
        .iter()
        .map(|mv| (game.move_to_san(mv), mv.to_lan()))
        .collect();
    legal_rows.sort();
    println!("{} legal moves:", legal_rows.len());
    for (san, lan) in legal_rows {
        println!("  {san:<6} {lan}");
    }

    let knight = Position::from_algebraic("g1")?;
    let knight_moves = game.legal_moves_for_position(&knight);
    let mut knight_rows: Vec<_> = knight_moves
        .iter()
        .map(|mv| (game.move_to_san(mv), mv.to_lan()))
        .collect();
    knight_rows.sort();
    println!();
    println!("Moves from g1:");
    for (san, lan) in knight_rows {
        println!("  {san:<6} {lan}");
    }

    Ok(())
}
